<a href="https://colab.research.google.com/github/B-i-g-g-i-e/Data-Visualization-Students-Tasks/blob/main/Sprints.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Student Performance and Learning Behavior

SPRINT1

MILESTONE_1

Data Acquisition:
The following data was acquired from
https://www.kaggle.com/datasets/muhammadazam121/student-performance-data

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('Student_performance_data.csv')

#prints first 5 rows
display(df.head())


,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
0,1001,17,1,0,2,19.833723,7,1,2,0,0,1,0,2.929196,2.0
1,1002,18,0,0,1,15.408756,0,0,1,0,0,0,0,3.042915,1.0
2,1003,15,0,2,3,4.210570,26,0,2,0,0,0,0,0.112602,4.0
3,1004,17,1,0,3,10.028829,14,0,3,1,0,0,0,2.054218,3.0
4,1005,17,1,0,2,4.672495,17,1,3,0,0,0,0,1.288061,4.0


In [ ]:

# shows column names and data types
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          2392 non-null   int64  
 1   Age                2392 non-null   int64  
 2   Gender             2392 non-null   int64  
 3   Ethnicity          2392 non-null   int64  
 4   ParentalEducation  2392 non-null   int64  
 5   StudyTimeWeekly    2392 non-null   float64
 6   Absences           2392 non-null   int64  
 7   Tutoring           2392 non-null   int64  
 8   ParentalSupport    2392 non-null   int64  
 9   Extracurricular    2392 non-null   int64  
 10  Sports             2392 non-null   int64  
 11  Music              2392 non-null   int64  
 12  Volunteering       2392 non-null   int64  
 13  GPA                2392 non-null   float64
 14  GradeClass         2392 non-null   float64
dtypes: float64(3), int64(12)
memory usage: 280.4 KB


None

The data contains 2392 entries and 15 variables, the two main data types are float64 and int32

In [ ]:
#calculates the mean,std,max,min for each column
display(df.describe())



,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass
count,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000,2392.000000
mean,2196.500000,16.468645,0.510870,0.877508,1.746237,9.771992,14.541388,0.301421,2.122074,0.383361,0.303512,0.196906,0.157191,1.906186,2.983696
std,690.655244,1.123798,0.499986,1.028476,1.000411,5.652774,8.467417,0.458971,1.122813,0.486307,0.459870,0.397744,0.364057,0.915156,1.233908
min,1001.000000,15.000000,0.000000,0.000000,0.000000,0.001057,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,1598.750000,15.000000,0.000000,0.000000,1.000000,5.043079,7.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,1.174803,2.000000
50%,2196.500000,16.000000,1.000000,0.000000,2.000000,9.705363,15.000000,0.000000,2.000000,0.000000,0.000000,0.000000,0.000000,1.893393,4.000000
75%,2794.250000,17.000000,1.000000,2.000000,2.000000,14.408410,22.000000,1.000000,3.000000,1.000000,1.000000,0.000000,0.000000,2.622216,4.000000
max,3392.000000,18.000000,1.000000,3.000000,4.000000,19.978094,29.000000,1.000000,4.000000,1.000000,1.000000,1.000000,1.000000,4.000000,4.000000


The minimum age is 15 y.o and the maximum age is 18 y.o, hence the range is from 15-18 y.o
The average GPA is 1.9 while the spread from the average is 0.9

In [ ]:
#check for missing values
display(df.isnull().sum())

,0
StudentID,0
Age,0
Gender,0
Ethnicity,0
ParentalEducation,0
StudyTimeWeekly,0
Absences,0
Tutoring,0
ParentalSupport,0
Extracurricular,0


Data Quality:
This data has no missing values

MILESTONE_2

Data Cleaning: Because the dataset contained zero missing values, we focused data cleaning to ensuring gpa had only a maximum of 4.0

In [ ]:
#Checking for anomalies

weird_gpas = df[(df['GPA'] > 4.0) | (df['GPA'] < 0.0)]

# Print out how many weird GPAs we found
print(f"No. of weird GPAs found: {len(weird_gpas)}")



No. of weird GPAs found: 0


there were no weird anomalies in the data

In [ ]:
# in case of a number >4.0 or <0 clip() will force the number to 4 or 0 depending on its range
df['GPA'] = df['GPA'].clip(upper=4.0, lower=0.0)



In [ ]:
# check if it worked
print(f"New Max GPA: {df['GPA'].max()}")
print(f"New Min GPA: {df['GPA'].min()}")


New Max GPA: 4.0
New Min GPA: 0.0


In [ ]:
# Feature engineering
# creating new data from old data

df['Total_Activities'] = df['Extracurricular'] + df['Sports'] + df['Music'] + df['Volunteering']

print(df['Total_Activities'])


0       1
1       0
2       0
3       1
4       0
       ..
2387    1
2388    1
2389    1
2390    2
2391    1
Name: Total_Activities, Length: 2392, dtype: int64


In [ ]:
df['under_performing'] = np.where((df['Absences'] > 15) | (df['GPA'] < 2.0), 1, 0)
print(df['under_performing'])

0       0
1       0
2       1
3       0
4       1
       ..
2387    0
2388    0
2389    1
2390    1
2391    0
Name: under_performing, Length: 2392, dtype: int64


In [ ]:
#removing the floats and creating three portions to work with
df['Study_Habits'] = pd.qcut(df['StudyTimeWeekly'], q=3, labels=['Low', 'Medium', 'High'])

print(df['Study_Habits'])

0         High
1         High
2          Low
3       Medium
4          Low
         ...  
2387    Medium
2388    Medium
2389    Medium
2390    Medium
2391      High
Name: Study_Habits, Length: 2392, dtype: category
Categories (3, object): ['Low' < 'Medium' < 'High']


In [ ]:
display(df[['StudentID', 'Total_Activities', 'under_performing', 'Study_Habits']])

,StudentID,Total_Activities,under_performing,Study_Habits
0,1001,1,0,High
1,1002,0,0,High
2,1003,0,1,Low
3,1004,1,0,Medium
4,1005,0,1,Low
...,...,...,...,...
2387,3388,1,0,Medium
2388,3389,1,0,Medium
2389,3390,1,1,Medium
2390,3391,2,1,Medium


In [ ]:
df.to_csv('Processed_Student_Data.csv', index=False)

new_df = pd.read_csv('Processed_Student_Data.csv')

display(new_df.head(10))

,StudentID,Age,Gender,Ethnicity,ParentalEducation,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Extracurricular,Sports,Music,Volunteering,GPA,GradeClass,underperforming,under_performing,Study_Habits,Total_Activities
0,1001,17,1,0,2,19.833723,7,1,2,0,0,1,0,2.929196,2.0,0,0,High,1
1,1002,18,0,0,1,15.408756,0,0,1,0,0,0,0,3.042915,1.0,0,0,High,0
2,1003,15,0,2,3,4.210570,26,0,2,0,0,0,0,0.112602,4.0,1,1,Low,0
3,1004,17,1,0,3,10.028829,14,0,3,1,0,0,0,2.054218,3.0,0,0,Medium,1
4,1005,17,1,0,2,4.672495,17,1,3,0,0,0,0,1.288061,4.0,1,1,Low,0
5,1006,18,0,0,1,8.191219,0,0,1,1,0,0,0,3.084184,1.0,0,0,Medium,1
6,1007,15,0,1,1,15.601680,10,0,3,0,1,0,0,2.748237,2.0,0,0,High,1
7,1008,15,1,1,4,15.424496,22,1,1,1,0,0,0,1.360143,4.0,1,1,High,1
8,1009,17,0,0,0,4.562008,1,0,2,0,1,0,1,2.896819,2.0,0,0,Low,2
9,1010,16,1,0,1,18.444466,0,0,3,1,0,0,0,3.573474,0.0,0,0,High,1
